### PR TITLE
gh-117657: TSAN Fix races in `PyMember_Get` and `PyMember_Set`, for C extensions

### PR DESCRIPTION
**Pull request by dpdani**
Fix data races that would only be visible when using C extensions.

This is a follow-up on #119368.

I'm intentionally not testing:
- `_Py_T_NONE` (deprecated)
- `_Py_T_OBJECT` (deprecated)
- `Py_T_STRING` (immutable)
- `Py_T_STRING_INPLACE` (immutable)

#### Py_T_CHAR

For some reason `Py_T_CHAR` is untested also in `test_capi.test_structmembers`.
In fact, it's not even in the supporting C types, which I'm using for the TSAN tests as well: [old api](https://github.com/python/cpython/blob/ec89620e5e147ba028a46dd695ef073a72000b84/Modules/_testcapi/structmember.c#L119), [new api](https://github.com/python/cpython/blob/ec89620e5e147ba028a46dd695ef073a72000b84/Modules/_testcapi/structmember.c#L33).
I'm wondering if there's a specific reason for this, or if it should be in the test suite instead.

I'm guessing that the TSAN suite should cover the thread-safety of `Py_T_CHAR` regardless?
Or should we dismiss it for TSAN tests as well?

I've added a new member to the `T_CHAR` member to the `_testcapi` module, and the rest of the tests don't seem to break.
I can revert this change if needed.

* Issue: gh-117657

### CODE REVIEW

File: Include/cpython/pyatomic.h [hunk 1]
```diff
@@ -321,6 +321,27 @@ _Py_atomic_load_ptr(const void *obj);
 static inline int
 _Py_atomic_load_int_relaxed(const int *obj);
 
+static inline char
+_Py_atomic_load_char_relaxed(const char *obj);
```
Comment (picnixz) [inline Include/cpython/pyatomic.h:325]: I know this is not in the scope of this PR, but those additions make me feel that all this code should be auto-generated by some new clinic magic for instance. One reason is that it's easy to miss a type when C/C.

File: Include/internal/pycore_pyatomic_ft_wrappers.h [hunk 2]
```diff
@@ -61,14 +61,61 @@ extern "C" {
...
+#define FT_ATOMIC_LOAD_UINT_RELAXED(value) \
+    _Py_atomic_load_uint_relaxed(&value)
+#define FT_ATOMIC_STORE_LONG_RELEASE(value, new_value) \
+    _Py_atomic_store_long_release(&value, new_value)
+#define FT_ATOMIC_LOAD_LONG_RELAXED(value) \
+    _Py_atomic_load_long_relaxed(&value)
+#define FT_ATOMIC_STORE_ULONG(value, new_value) \
+    _Py_atomic_store_ulong(&value, new_value)
+#define FT_ATOMIC_STORE_SSIZE_RELEASE(value, new_value) \
+    _Py_atomic_store_ssize_release(&value, new_value)
```
Comment (picnixz) [inline Include/internal/pycore_pyatomic_ft_wrappers.h]: Those two are asymmetric in their declaration (namely, I cannot find the `_Py_atomic_load_ulong_relaxed` nor the `_Py_atomic_load_ssize_relaxed`). Are they somewhere else or just not needed? (or maybe behind a if-guard?)

**Review by picnixz — COMMENTED** [review]
Quite hard to review since everything looks the same and the functions are not ordered by pairs apparently. I didn't go through everything but I had a question and an observation.

File: Include/internal/pycore_pyatomic_ft_wrappers.h (hunk 2 above)
Comment (dpdani) [reply Include/internal/pycore_pyatomic_ft_wrappers.h]: They were already defined:
- [_Py_atomic_load_ulong_relaxed](https://github.com/python/cpython/blob/297f2e093ec95800ae2184330b8408c875523467/Include/cpython/pyatomic.h#L552)
- [_Py_atomic_load_ssize_relaxed](https://github.com/python/cpython/blob/297f2e093ec95800ae2184330b8408c875523467/Include/cpython/pyatomic.h#L358)

File: Include/cpython/pyatomic.h (hunk 1 above)
Comment (dpdani) [reply Include/cpython/pyatomic.h:325]: Totally agree.

It's exactly what happened here: https://github.com/python/cpython/pull/123211/commits/50a7bb61adc5c6f00f627514f6729f20289778c7

File: Include/internal/pycore_pyatomic_ft_wrappers.h (hunk 2 above)
Comment (picnixz) [reply Include/internal/pycore_pyatomic_ft_wrappers.h]: Huh it's really hard to review when the pieces are not at the correct places (I know that it's better for your PR not to change the ordering in this module but I honestly think we need to do one PR which groups correctly the macros and functions...)

**Comment by dpdani** [comment]
I'm not exactly sure why `test_importlib` fails, but I don't want to push a dummy commit just to retry it.

Anyways, the TSAN checks for `test_free_threading.test_slots` are 🟢

**Comment by picnixz** [comment]
> I'm not exactly sure why test_importlib fails, but I don't want to push a dummy commit just to retry it.

I've relaunched the test manually.

File: Include/internal/pycore_pyatomic_ft_wrappers.h (hunk 2 above)
Comment (dpdani) [reply Include/internal/pycore_pyatomic_ft_wrappers.h]: Came very close to doing that today

File: Python/structmember.c [hunk 3]
```diff
@@ -169,16 +172,16 @@ PyMember_SetOne(char *addr, PyMemberDef *l, PyObject *v)
             return -1;
         }
         if (v == Py_True)
-            *(char*)addr = (char) 1;
+            FT_ATOMIC_STORE_CHAR_RELEASE(*(char*)addr, 1);
```
Comment (colesbury) [inline Python/structmember.c]: The stores should all be "relaxed" ordering here.

File: Python/structmember.c [hunk 4]
```diff
@@ -243,15 +246,15 @@ PyMember_SetOne(char *addr, PyMemberDef *l, PyObject *v)
...
             if (ulong_val > UINT_MAX) {
                 WARN("Truncation of value to unsigned int");
             }
         }
         break;
     }
     case Py_T_LONG:{
-        *(long*)addr = PyLong_AsLong(v);
+        FT_ATOMIC_STORE_LONG_RELEASE(*(long*)addr, PyLong_AsLong(v));
         if ((*(long*)addr == -1) && PyErr_Occurred())
```
Comment (colesbury) [inline Python/structmember.c]: This pattern is broken -- the error check needs to happen before the store to `addr`.

There are a few more instances of this in this file.

File: Python/structmember.c [hunk 5]
```diff
@@ -80,7 +81,9 @@ PyMember_GetOne(const char *obj_addr, PyMemberDef *l)
         v = PyUnicode_FromString((char*)addr);
         break;
     case Py_T_CHAR:
-        v = PyUnicode_FromStringAndSize((char*)addr, 1);
+        // case ...: char value = ... requires a C23-compatible compiler
+        value = FT_ATOMIC_LOAD_CHAR_RELAXED(*(char *) addr);
+        v = PyUnicode_FromStringAndSize(&value, 1);
```
Comment (colesbury) [inline Python/structmember.c]: If you want a local variable in a case statement, declare a scope:

```
case Py_T_CHAR: {
    char value = FT_ATOMIC_LOAD_CHAR_RELAXED(*(char *) addr);
    v = PyUnicode_FromStringAndSize(&value, 1);
}
```

**Review by colesbury — COMMENTED** [review]
Thanks. I left a few comments below.

How long do the added tests takes to run?

File: Include/internal/pycore_pyatomic_ft_wrappers.h (hunk 2 above)
Comment (picnixz) [reply Include/internal/pycore_pyatomic_ft_wrappers.h]: FTR, I opened #123385 where macros are being used. The code looks better IMO (it also reduces 1500 lines to 500 lines). [idea was dropped, see the PR for the rationale]

File: Python/structmember.c (hunk 5 above)
Comment (dpdani) [reply Python/structmember.c]: Oh didn't know that, thanks!

File: Python/structmember.c (hunk 4 above)
Comment (dpdani) [reply Python/structmember.c]: Yeah, I saw that, but thought not to change the behavior.
I'll fix that.

I guess we would need a news entry then.

**Comment by dpdani** [comment]
> How long do the added tests takes to run?

```
./python -m test test_free_threading.test_slots

...
Total duration: 2.3 sec
...
```

**Comment by dpdani** [comment]
@colesbury can you take another look?

File: Python/structmember.c [hunk 6]
```diff
@@ -79,9 +79,11 @@ PyMember_GetOne(const char *obj_addr, PyMemberDef *l)
     case Py_T_STRING_INPLACE:
         v = PyUnicode_FromString((char*)addr);
         break;
-    case Py_T_CHAR:
-        v = PyUnicode_FromStringAndSize((char*)addr, 1);
+    case Py_T_CHAR: {
+        char value = 0;
```
Comment (colesbury) [inline Python/structmember.c]: This is missing the read from `addr`

File: Python/structmember.c [hunk 7]
```diff
@@ -278,12 +280,12 @@ PyMember_SetOne(char *addr, PyMemberDef *l, PyObject *v)
...
                 return -1;
             }
-            *(unsigned long*)addr = ulong_val;
+            FT_ATOMIC_STORE_ULONG_RELAXED(*(unsigned long *)addr, ulong_val);
         }
         break;
     }
     case Py_T_PYSSIZET:{
-        *(Py_ssize_t*)addr = PyLong_AsSsize_t(v);
+        FT_ATOMIC_STORE_SSIZE_RELAXED(*(Py_ssize_t*)addr, PyLong_AsSsize_t(v));
```
Comment (colesbury) [inline Python/structmember.c]: The store should be moved after the error check

File: Python/structmember.c [hunk 8]
```diff
@@ -293,11 +295,11 @@ PyMember_SetOne(char *addr, PyMemberDef *l, PyObject *v)
         double double_val = PyFloat_AsDouble(v);
         if ((double_val == -1) && PyErr_Occurred())
             return -1;
-        *(float*)addr = (float)double_val;
+        FT_ATOMIC_STORE_FLOAT_RELAXED(*(float*)addr, (float)double_val);
         break;
         }
     case Py_T_DOUBLE:
-        *(double*)addr = PyFloat_AsDouble(v);
+        FT_ATOMIC_STORE_DOUBLE_RELAXED(*(double*)addr, PyFloat_AsDouble(v));
```
Comment (colesbury) [inline Python/structmember.c]: The store should be moved after the error check

File: Python/structmember.c [hunk 9]
```diff
@@ -318,16 +320,16 @@ PyMember_SetOne(char *addr, PyMemberDef *l, PyObject *v)
...
         }
     case Py_T_STRING:
     case Py_T_STRING_INPLACE:
         PyErr_SetString(PyExc_TypeError, "readonly attribute");
         return -1;
     case Py_T_LONGLONG:{
-        long long value;
-        *(long long*)addr = value = PyLong_AsLongLong(v);
+        long long value = PyLong_AsLongLong(v);
+        FT_ATOMIC_STORE_LLONG_RELAXED(*(long long*)addr, value);
```
Comment (colesbury) [inline Python/structmember.c]: The store should be moved after the error check

File: Include/cpython/pyatomic_msc.h [hunk 10]
```diff
@@ -899,6 +947,60 @@ _Py_atomic_store_ullong_relaxed(unsigned long long *obj,
...
+static inline void
+_Py_atomic_store_char_relaxed(char *obj, char value)
+{
+    *(volatile char *)obj = value;
+}
+
+static inline void
+_Py_atomic_store_uchar_relaxed(unsigned char *obj, unsigned char value)
+{
+    *(volatile unsigned char*)obj = value;
```
Comment (colesbury) [inline Include/cpython/pyatomic_msc.h]:
```suggestion
    *(volatile unsigned char *)obj = value;
```

File: Include/cpython/pyatomic_msc.h [hunk 11]
```diff
@@ -899,6 +947,60 @@ _Py_atomic_store_ullong_relaxed(unsigned long long *obj,
...
+}
+
+static inline void
+_Py_atomic_store_short_relaxed(short *obj, short value)
+{
+    *(volatile short*)obj = value;
```
Comment (colesbury) [inline Include/cpython/pyatomic_msc.h]:
```suggestion
    *(volatile short *)obj = value;
```

File: Include/cpython/pyatomic_msc.h [hunk 12]
```diff
@@ -899,6 +947,60 @@ _Py_atomic_store_ullong_relaxed(unsigned long long *obj,
...
+}
+
+static inline void
+_Py_atomic_store_ushort_relaxed(unsigned short *obj, unsigned short value)
+{
+    *(volatile unsigned short*)obj = value;
```
Comment (colesbury) [inline Include/cpython/pyatomic_msc.h]:
```suggestion
    *(volatile unsigned short *)obj = value;
```

File: Include/cpython/pyatomic_msc.h [hunk 13]
```diff
@@ -899,6 +947,60 @@ _Py_atomic_store_ullong_relaxed(unsigned long long *obj,
...
+}
+
+static inline void
+_Py_atomic_store_uint_release(unsigned int *obj, unsigned int value)
+{
+    *(volatile unsigned int*)obj = value;
```
Comment (colesbury) [inline Include/cpython/pyatomic_msc.h]:
```suggestion
    *(volatile unsigned int *)obj = value;
```

File: Include/cpython/pyatomic_msc.h [hunk 14]
```diff
@@ -899,6 +947,60 @@ _Py_atomic_store_ullong_relaxed(unsigned long long *obj,
...
+static inline void
+_Py_atomic_store_double_relaxed(double *obj, double value)
+{
+    *(volatile double *)obj = value;
+}
+
+static inline void
+_Py_atomic_store_llong_relaxed(long long *obj, long long value)
+{
+    *(volatile long long*)obj = value;
```
Comment (colesbury) [inline Include/cpython/pyatomic_msc.h]:
```suggestion
    *(volatile long long *)obj = value;
```

**Review by colesbury — COMMENTED** [review]
This mostly looks good. Please add a `char` test to `test_structmembers.py`. I think the implementation in the PR is broken (see inline comment) and it's not caught by any unit test.

It's fine to ping me on PRs like you did here, but please also use GitHub's "re-request review" button (top-right corner, in the reviewers section). It means that the PR shows up in my list of "awaiting reviews".

**Comment by dpdani** [comment]
Sorry, will do.

**Comment by dpdani** [comment]
I finally had some time to come back to this.

I guess that moving the stores after error checking does slightly change the externally-visible behavior.
Should we write a news entry?

**Review by colesbury — APPROVED** [review]
LGTM

**Comment by colesbury** [comment]
!buildbot nogil

**Comment by colesbury** [comment]
Thanks @dpdani!

**Comment by dpdani** [comment]
that sounds like my fault, I'll take a look later today

**Comment by colesbury** [comment]
Let's skip the test on s390x for now while we investigate.

I'm not sure what's causing the failure. The only thing I can think of is that `char` is unsigned on s390x (vs. signed on GCC x86-64 and arm64).

**Comment by colesbury** [comment]
https://github.com/python/cpython/pull/127577 should fix it. The problem was that the 'C' format code expects an `int` as the destination (not a `char`):

https://docs.python.org/3/c-api/arg.html#c.Py_BuildValue

This happens to work out okay on little endian systems because the non-zero part of the value is at the same place (the first byte), and it just writes some zeros to the struct's padding bytes. On big endian systems, like s390x, it would write zero to `char_member` and some non-zero value to parts of the padding bytes.

I followed the instructions on https://docs.gitlab.com/omnibus/development/s390x.html to debug s390x using Docker and QEMU. I had to slightly modify their commands: in particular, I had to specify `--platform linux/s390x`:

```
docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
docker run --platform linux/s390x -v $(pwd):/cpython --rm -it s390x/ubuntu
```

**Comment by dpdani** [comment]
oops, my bad

thank you for all the details 🙏

**Pull request merged**